Refactor endpoint and policy packages to separate test code from production code

roykharman opened the pull request:
This pull request refactors the policy and endpoint packages to improve code clarity, reduce unnecessary complexity, and ensure clear distinction between production and test-specific code, addressing issue #34069.

Key Changes:

[1] Removed Unused testing.TB Parameters:
In the l4.go file, the unused testing.TB parameters were removed from test-related functions. This reduces unnecessary complexity and ensures these functions are cleaner and not tied to the testing package when it's not needed.

[2] Renamed Test-Specific Functions:
Renamed the Equals and Diff functions in l4.go to TestingOnlyEquals and TestingOnlyDiff, respectively. This change explicitly marks these functions as intended for testing purposes only, helping to prevent accidental use in production code.

[3] Clarifying Note in NewTestEndpointWithState:
Added a comment to the NewTestEndpointWithState function in the endpoint package, clearly indicating that this function is intended solely for testing purposes. This helps future developers quickly understand the function's purpose.

[4] Linting and Error Fixes:
Fixed several linting errors and addressed undefined references caused by the initial renaming of test-specific functions. All necessary functions were renamed appropriately, and the code passes linting checks.

Challenges Faced:
While we initially aimed to refactor parts of the kvstore package, the decision was ultimately made to focus on renaming functions and removing unnecessary parameters in the policy and endpoint packages. The final solution ensures that test-specific functions are clearly marked, simplifying future maintenance and reducing the risk of test code being accidentally used in production environments.

Regarding the kvstore package we attempted to move test-specific code into separate test files to better isolate testing logic from production code. However, this approach caused several undefined reference errors between packages, complicating access to internal functions. We also considered moving the test logic to utility files, but this did not resolve the accessibility issues.

Towards: #34069
Signed-off-by: Roy roykharman@gmail.com
Reviewed-by: Sameeh Jubran sameeh.j@gmail.com

```release-note
Refactored the endpoint and policy packages to separate test-specific code from production code.
```

roykharman committed:
Refactor NewTestEndpointWithState: Remove unused testing parameter (#34069)

- Removed an unused `testing.TB` parameter from the `NewTestEndpointWithState` function in `endpoint.go`.
- This parameter was not being utilized within the function, and its removal simplifies the function signature.
- The `testing.TB` parameter seems redundant in this context, as the function does not rely on it for any testing behavior.
- Removing this parameter also eliminates the need for the `testing` package import in production code, contributing to the resolution of issue #34069 by reducing the unnecessary weight and surface area of the production build.

Signed-off-by: Roy <roykharman@gmail.com>
Reviewed-by: Sameeh Jubran <sameeh.j@gmail.com>

roykharman committed:
Refactor l4.go: Remove unused testing parameter (#34069)

- Removed an unused `testing.TB` parameter from a function in `l4.go`.
- This parameter was not being used within the function and was likely a leftover placeholder.
- By removing this unused parameter, the function is simplified and no longer depends on the `testing` package.
- This change contributes to resolving issue #34069 by removing the unnecessary inclusion of the Go built-in "testing" package from production builds, reducing both weight and surface area.

Signed-off-by: Roy <roykharman@gmail.com>
Reviewed-by: Sameeh Jubran <sameeh.j@gmail.com>

giorio94 commented on pkg/endpoint/endpoint.go:
```diff
@@ -15,7 +15,6 @@ import (
 	"strconv"
 	"strings"
 	"sync/atomic"
-	"testing"
```
/cc @aanm who authored f6def942a5e73ad9356e9a0964740649ee2b109f

giorio94 commented on pkg/policy/l4.go:
```diff
@@ -12,7 +12,6 @@ import (
 	"strconv"
 	"strings"
 	"sync/atomic"
-	"testing"
```
/cc @nathanjsweet who authored d63157d0ad02b

giorio94 commented on pkg/kvstore/kvstoreTest/dummy.go:
Nit. Package names should not contain uppercase letters: https://google.github.io/styleguide/go/decisions.html#package-names

giorio94 commented on pkg/kvstore/etcd_test.go:
```diff
@@ -14,13 +14,91 @@ import (
+// in case different tests are run in parallel. A cleanup function is
+// automatically registered to delete all keys and close the client when the
+// test terminates.
+func SetupDummyWithConfigOpts(tb testing.TB, dummyBackend string, opts map[string]string) {
```
This is duplicating a non-trivial function in two different places, which will eventually diverge. I'd suggest trying to refactor the structure to avoid requiring this duplication.

giorio94 commented on pkg/kvstore/export.go:
```diff
@@ -0,0 +1,45 @@
+	e.config.Endpoints = []string{etcdDummyAddress}
+}
+
+func ExportedGetBackend(name string) backendModule {
```
The fact that a given variable/method is exported is already implied by the name starting with an uppercase latter. No need to explicitly name it `Exported`. I do also see limited value in creating an alias name of an already existing function, it should be just exported if it needs to be so.

giorio94 commented on pkg/kvstore/export.go:
```diff
@@ -0,0 +1,45 @@
+
+func NewIncTimer() (inctimer.IncTimer, func() bool) {
+	return inctimer.New()
+}
```
Why is this alias needed?

giorio94 commented on pkg/kvstore/etcd_test.go:
```diff
@@ -14,13 +14,91 @@ import (
+		if err != nil {
+			tb.Fatalf("Unable to set config options for kvstore backend module: %v", err)
+		}
+	}
```
I'd suggest trying to rework this snippet using `kvstore.NewClient` (or `kvstore.Setup`), which is already exported, and should avoid the need for exporting a bunch of dedicated test function (which may be even deleted afterwards).

giorio94 commented on pkg/kvstore/kvstoreTest/dummy.go:
```diff
@@ -33,36 +27,48 @@ func SetupDummy(tb testing.TB, dummyBackend string) {
+	// Ensure the module is valid
+	if module == nil {
+		tb.Fatalf("Unknown dummy kvstore backend %s", dummyBackend)
+	}
```
This code is duplicated.

giorio94 reviewed (changes requested):
Thanks @roykharman for taking care of this.

The first two commits look pretty straightforward to me. I've additionally cced the original authors who had introduced the usage of testing.T in these methods, to get their opinion as well. Based on the original commit messages, the idea there was to make it clear that the given methods are intended for testing purposes only, and to prevent using them in production code.

I'm a bit more doubtful about the kvstore changes. Although I agree with the overall intent, the current changes are leading to duplicated code (which will deverge over time), and the need of exposing methods that were intended for testing only. I've left a few suggestions on how to try to refactor those to simplify the logic and replace the usage of some internal methods, but I'm not totally sure if it will be possible to achieve a really good solution without a significant refactoring of the kvstore package. Similarly, I'm also not really sure whether avoiding to import the testing package in production code really outweighs the extra complexity introduced here, but I guess that will become clearer after addressing the initial suggestions.

Please also note that the PR currently contains a merge commit, which are not allowed in Cilium. Please rebase your PR onto the main branch to fix the conflicts and remove it.

nathanjsweet commented on pkg/policy/l4.go:
```diff
@@ -1144,8 +1143,8 @@ type L4PolicyMap interface {
 	ForEach(func(l4 *L4Filter) bool)
-	Equals(t *testing.T, bMap L4PolicyMap) bool
-	Diff(t *testing.T, expectedMap L4PolicyMap) string
+	Equals(bMap L4PolicyMap) bool
```
The reason the 'testing.T` parameter is added to these methods is to ensure they are not used in production code. There, at least, be a convention that preserves this intent.
```suggestion
	TestingOnlyEquals(bMap L4PolicyMap) bool
```

nathanjsweet commented on pkg/policy/l4.go:
```diff
@@ -1144,8 +1143,8 @@ type L4PolicyMap interface {
-	Equals(t *testing.T, bMap L4PolicyMap) bool
-	Diff(t *testing.T, expectedMap L4PolicyMap) string
+	Equals(bMap L4PolicyMap) bool
+	Diff(expectedMap L4PolicyMap) string
```
Ditto above.
```suggestion
	TestingOnlyDiff(expectedMap L4PolicyMap) string
```

roykharman replied:
Thank you for the feedback, @nathanjsweet!

I've renamed the functions as you suggested to TestingOnlyEquals and TestingOnlyDiff to preserve the intent of restricting their usage to testing scenarios, I have also added a note in endpoint.go above the NewTestEndpointWithState function the it should be used for testing purposes only since I also removed "testing" from this file. 
Please let me know if there's anything else you'd like me to address.

nathanjsweet reviewed (changes requested):
Thank you so much @roykharman  !
Some small nits.

roykharman committed:
Refactor: Rename functions in l4.go and add testing note (#34069)

- Renamed two functions in l4.go to 'TestingOnly*' prefix to clarify their use for testing only.
- Added a note to NewTestEndpointWithState indicating it is for testing purposes.
- This change contributes to resolving issue #34069 by making test-specific code more explicit.

Signed-off-by: Roy <roykharman@gmail.com>
Reviewed-by: Sameeh Jubran <sameeh.j@gmail.com>

aanm reviewed (approved):
:+1: for the comment that Marco ping me about.

roykharman committed:
Fix remaining function name issues and resolve linting errors (#34069)

- Renamed remaining Equals and Diff functions to TestingOnlyEquals and TestingOnlyDiff.
- Fixed all linting errors.

Signed-off-by: Roy <roykharman@gmail.com>

giorio94 reviewed (approved):
Thanks!
